### ISSUE TITLE
Waiting for GPU is killing performance

### ISSUE THREAD

**Issue by hardcoremore** [Gamua/Adobe-Runtime-Support#307]
Hi @ajwfrost,

This is a huge issue Adobe never fixed that is destroying performance for NO GOOD reason. This seams to be happening on iOS, Windows and Android. I was not able to reproduce it with Max OS.

Can you please put this on a high priority, because it is destroying everything I worked for?

Here you can find Adobe Scout profile that I have just recorded: [http://tiny.cc/2tuunz](http://tiny.cc/2tuunz)

If you look at the profile you can see that my game is running smooth except when this Waiting for GPU kicks in for no good reason. You can observe this on the frames at the end of scout profile.

Please take a look at this, it is a really big issue for games.

Regards,

Caslav

**Comment by itlancer** [comment]
Related issues:
https://github.com/Gamua/Adobe-Runtime-Support/issues/85
https://tracker.adobe.com/#/view/AIR-4198654
https://tracker.adobe.com/#/view/AIR-3651843

**Comment by ajwfrost** [comment]
Hi

As far as I know, "waiting for GPU" is where we've submitted the requests (normally asynchronous) into the GPU and are then waiting for them to complete i.e. once we've completed a frame and are waiting before we can start the next frame rendering. But if this sort of thing is happening across platforms and mechanisms - Android/iOS using OpenGL ES, Windows using Direct3D 11 - then it may imply there's something that the runtime is doing here which is wrong.

Could we get a copy of this game to run this locally and see if we can reproduce it on the development machines - are there any special steps to trigger this (there are lots of mouse events then key presses happening.. or do we just play the game!)

thanks

**Comment by hardcoremore** [comment]
Hi @ajwfrost,

Where I can send you the game?

I have a .exe file of my game which is not released yet and has test levels with which I can reproduce this every time. It always happen in under a minute of gameplay.

**Comment by ajwfrost** [comment]
Can you try uploading with the below please?
https://transfer.harman.com/requests/aQ4rFkPgIETfoiPrOuqOxU

**Comment by hardcoremore** [comment]
Hi @ajwfrost ,

I have sent you release build of the game. Please confirm if you receive all the files and if you can debug it.

Regards,

Caslav

**Comment by ajwfrost** [comment]
Thanks @hardcoremore, yes we've got the files and will look into it. Curious about this not happening when on battery, as far as I'm aware AIR wouldn't be doing anything differently based on the power state!

But yes, we can try to see what's actually happening regarding all the timings, and what's making it wait..

**Comment by hardcoremore** [comment]
Hi Andrew,

Any news on this. Where you able to reproduce the issue with the app I sent you?

Thanks,
Caslav

**Comment by hardcoremore** [comment]
Hi @ajwfrost,

Any news on this one? Where you able to reproduce issue with the game I have sent you.

Regards,
Caslav

**Comment by ajwfrost** [comment]
Hi - sorry, missed that earlier request! Will check on this one, I have a recollection that we did reproduce it with your app but hadn't got further with it.. but I'll need to double-check with the team next week.

thanks

**Comment by hardcoremore** [comment]
Thanks @ajwfrost,

Its great that you were able to reproduce this issue with the game I sent you.

Looking forward to fixing this ugly bug because it is a really big blocker for games.


Regards,
Caslav

**Comment by hardcoremore** [comment]
Hi @ajwfrost,

Was there any progress on this issue?

Regards,
Caslav

**Comment by ajwfrost** [comment]
@hardcoremore not got anywhere with this as yet, aiming to understand that GPU memory issue first as the two may be linked (typically if we settled into a period where there's a need to do extra GC clean-up, this sort of drop in performance could be a result...)

**Comment by ajwfrost** [comment]
@hardcoremore  sorry we can't find the game that you sent us for some reason .. is it possible to get that again as the online cache has expired? I'd like to check it against the runtime here that's fixed the GPU memory leak..
https://transfer.harman.com/requests/lAvvcpQgLDixUe1A90EZ5B

thanks

**Comment by hardcoremore** [comment]
Hi @ajwfrost,

Awesome, I have just sent the new version. Hope to see this bug fixed as well.

**Comment by hardcoremore** [comment]
Also @ajwfrost, I have one more request.

Can you please load level 42 "Polygon Collision Test" from the game I have sent you and play it. 

It will be slow because there are a lot of calculations and function calls (physics simulation) but can you please do some quick profiling just to check if there is something that can improve performance that Adobe didn't implement right or bother to do.

Just to check if something stands out that can be implemented in much more efficient way. It would mean a lot and I think that can be very good real scenario to stress test AIR.


Thanks

**Comment by hardcoremore** [comment]
Hi @ajwfrost,


Any news about this issue?

**Comment by hardcoremore** [comment]
I have tested this with latest AIR 33.1.1.217 and I can still reproduce it always. It destroys performance on PC completely by locking it to 30FPS.

**Comment by Adolio** [comment]
Hi everyone,

I think I'm now facing this issue as well. My game is usually running smoothly at 60 fps but frequently the frame rate falls down to 30 fps for no apparent reasons and gets stuck to this value until I do some (big?) changes in the rendering (like showing the pause overlay panel or resetting the current level).
That's a bit weird and unfortunately I don't have anything on my side to easily reproduce the issue...

I'm using Starling 2.6 & the latest Adobe AIR SDK from Harman (31.1.1.217) under Windows 10 (with a 64-bit build).

Best,
Aurélien

**Comment by Seitei** [comment]
I can confirm same happens to me, after a Windows update last night, the debugger version runs at a capped ~32 FPS.

**Comment by tudal** [comment]
I have the same issue. Only in debug mode. Weirdly when I have open Scout with ActionScript Sampler enabled then I have 60 fps back. Please fix this because testing on debug is pain right now.

**Comment by OliverJoyce** [comment]
I can confirm that since the latest Windows 10 update, this is happening to me too. AIR apps that ran at 60fps are now capped at around 32fps. This is a real showstopping bug for AIR, anyone found a solution yet?

**Comment by hardcoremore** [comment]
Hi @ajwfrost, is there any progress with this issue?

**Comment by teotigraphix** [comment]
Same, the Windows Update just capped by frame rate to 30 as well. Guess I will try and go back to an older SDK so I can test for "real".

**Comment by teotigraphix** [comment]
UPDATE: I have managed to hack fix this;

![image](https://user-images.githubusercontent.com/226417/96018897-ca8ad700-0e19-11eb-8fce-1c1c05ddb944.png)

I have a feeling it's the Power Management mode which was at `Optimum`, I set it back to max performance.

I also changed the max FPS but I don't think that was it, the same issue happens with USB and performance audio, if the settings get messed up, you get crackling, here we get clamped frame rate.

I have run the debugger multiple times and I finally get 60 FPS back on Windows 10.

**Comment by BMGMobile** [comment]
Sharing a workaround from another issue. It appears that this bug only occurs if the frame rate is set near or lower than the display refresh rate. Doing the following seems to work around the issue:
`stage.frameRate = 1000.0;`

If you're using the `ENTER_FRAME` callback you will be called after each vsync, so 60 FPS on a typical monitor, but more frequently on 120/144/240/320Hz monitors. You'll have to manage your own logic to handle faster refresh rates.

**Comment by ajwfrost** [comment]
Belatedly picking up this thread.. it's looking possible that the fact that things slow down after a while could be due to the GPU getting hot and therefore getting throttled, which would perhaps tie in with some of the observations (and hack for changing the power management mode). But also if it's something that would be based on some clock skews based on framerates, then we may want to change how the enter-frame thing works a little.

Currently you set the desired frame rate e.g. via the authoring tools, which is then embedded into the SWF file header. Then there's some logic in the core rendering loop that decides whether or not it's time to draw a new frame, based on the time since the last one (it tries to catch up if the actual frame processing/rendering takes too long). The trigger for this check, at least on Windows, is a timer that's going off at a higher rate (10x) where we're also checking for other activity such as input events and network updates etc.

So, if we come out of a rendering loop and then expect to have a frame delay of 1000/60 = 16ms, it might be that by the time we've done the next bit of rendering we have then missed that next vsync slot and we just get delayed due to the way in which the GPU works. i.e.
Time 1000 -> Frame 1 start processing
Time 1010 -> Frame 1 complete processing
Time 1016 -> Flip GPU so frame appears
Time 1022 perhaps -> Frame 2 start processing
Time 1032 -> Flip GPU but nothing provided yet
Time 1032 -> Frame 2 complete processing
Time 1048 -> Flip GPU so frame appears

This is a bit of conjecture but we'll do some checking to see if this might be what's happening. We've been trying to work out where Scout gets its numbers from, but it appears this is doing some calculations based on the data that's getting dispatched from the AIR runtime, there's no "waiting for GPU" value being passed across which is making it a little harder to see where this is happening!

thanks

**Comment by BMGMobile** [comment]
Glad you're looking into it! From what we've seen, this is not a GPU throttling issue, especially since a workaround is to ask Adobe Air to render at 1000FPS. Our first reports of this issue were from players using 3rd party utilities to monitor frame rate, and we've been able to verify that our game stutters in this state, so it's not just a Adobe Scout issue.

**Comment by hardcoremore** [comment]
Thanks for looking at this Andrew!

**Comment by ajwfrost** [comment]
Hi all. There's some curious behaviour here ... but essentially it appears that:
1) enabling the advanced telemetry, or getting the player to do some other work, is helping in some way to keep the runtime working and running more smoothly - we're not sure why this it yet!
2) if you don't have much going on then after a while the Windows system seems to just call back our application less frequently which means we're then skipping frames...
3) ... and that's mostly down to how the main run loop is designed. It looks fine in theory, but in practice it doesn't seem to work how we would expect.

Actions we're looking at:
a) investigate further to see if there is something we can learn from the advanced telemetry setting or similar, to try to keep the frame rate up without having to do extra work!
b) reworking the main event loop -> could be a relatively tricky thing to do especially given how AIR can end up with multiple Windows all being controlled by one instance of the AVM - does anyone here actually use multiple NativeWindow objects as we could perhaps create a new mode for high-performance requirements at the expense of losing multi-window support?
c) actually have the internal timing logic change based on the `Stage.vsyncEnabled` setting (currently turning it off just means that D3D11 would do the 'present' and not wait anywhere, so framerate would be ignored?)
d) adding some other options into the AIR descriptor file so that you can choose some more control over things such as Direct3D level 

Comments/thoughts welcome..

thanks

**Comment by itlancer** [comment]
Hi, @ajwfrost 

> a) investigate further to see if there is something we can learn from the advanced telemetry setting or similar, to try to keep the frame rate up without having to do extra work!

Seems it related to https://github.com/Gamua/Adobe-Runtime-Support/issues/85

> b) reworking the main event loop -> could be a relatively tricky thing to do especially given how AIR can end up with multiple Windows all being controlled by one instance of the AVM - does anyone here actually use multiple NativeWindow objects as we could perhaps create a new mode for high-performance requirements at the expense of losing multi-window support?

In my observation the main issue somewhere when video (`Video` or `StageVideo`) or complex animations playback (or when high resolution used). Visual lags looks like "render waits something" and can not get in time. Also we use multiple `NativeWindow` objects in some of our applications (but lags reproduced even with only one `NativeWindow`). Moreover our apps decide it (if need to create another one `NativeWindow`) in run-time, so here some new option cannot be defined at AIR descriptor in our cases.
But this area definitely need to be fixed somehow. May be "new mode" that works only for main `NativeWindow` at first steps. May be use different AVMs or renderers for different windows.
Also working with multi-screens environment need to be improved. Right now there are some issues. For example: https://github.com/Gamua/Adobe-Runtime-Support/issues/309

> c) actually have the internal timing logic change based on the Stage.vsyncEnabled setting (currently turning it off just means that D3D11 would do the 'present' and not wait anywhere, so framerate would be ignored?)

In my observations changing `Stage.vsyncEnabled` value (at least for Windows devices in our cases) do nothing. Visually no benefits from it.
Note: for these test I don't use `Stage3D`, only `DisplayList` and `Video`/`StageVideo`.
May be I do something wrong or it can be applied only for `Stage3D`?
Also it could be useful for AIR to use latest DirectX 12 too: https://github.com/Gamua/Adobe-Runtime-Support/issues/502

> d) adding some other options into the AIR descriptor file so that you can choose some more control over things such as Direct3D level

At the first steps such approach great. If further such values will need to be changed at run-time we let you know!

**Comment by Zageron** [comment]
Hello, Flash Flash Revolution is definitely one of the greatest sufferers of these issues as we are a game that relies on extreme precision on the frame time. We would really appreciate an update on this.

We would also really like the ability to independently pump the event loop for keyboard input exclusively of the air runtime's callback system so we can acquire sub millisecond precision, or at least 1-2ms precision accuracy of keyboard presses to pass to the game.

If there is a known way to run an input loop completely independent of the rendering loop already that would be awesome to learn.

Currently we have extremely inconsistent performance and behaviour across almost every machine our game runs on, regardless of how powerful or weak machines are. (It is not consistent with the power of a machine, and seems to be more related to air not correctly reporting to the cpu and gpu that it would like the computer to be running at high performance.

Note: we do not use Harman because we are and OSS no profit and cannot afford the licencing for shipping let alone an arbitrary number of seats.

Thanks!

https://www.flashflashrevolution.com/

**Comment by Zageron** [comment]
:(

**Comment by Adolio** [comment]
@ajwfrost Following the recent changes made on `AIR 50.0.1.1` regarding the "performance issues if running without advanced telemetry connection", do you think we could expect any performance improvements in `release` builds too?

**Comment by ajwfrost** [comment]
Hi - that's a possibility perhaps, although the report under #85 didn't mention this being a problem in release builds. So the issue we were investigating there was that advanced telemetry was making the performance smoother. Oddly this was only an issue with Direct3D 11 it seems.. but the fix was a more generic one, which essentially improves the timer accuracy when we're asking to have a bit of down-time between frames. There was code in there which did this already, but only for certain (older) Windows versions. It looked like Adobe had seen in the MSFT documentation that this wasn't required any more, but it does actually seem to be needed!

If you have a reasonably reproducible way of checking this, then it's worth testing using 50.0.1.1 for sure, as the generic nature of the fix may have also improved that. It all goes back to the way that the animation/timer loops are set up in AIR, we'd looked into that previously (with a view to completely reworking the architecture) but not managed to get a solution; it would be great if this fix also impacts this one!

I've just re-read some of the above thread, it's very possible that we're dealing with multiple issues here though. There appears to have been a Windows update around August 2020 which changed things. But #85 was raised way before that... so very likely, this could improve things for some people/apps and not others :-(

Sorry, not a very clear/definitive answer here, "maybe" is probably the best summary!

**Comment by Adolio** [comment]
Thanks a lot for your answer Andrew. That's indeed pretty weird that this issue was touching mostly Direct3D 11.. Anyway, happy to see that you found a generic fix 👍

Unfortunately I don't have a specific reproducible case, I just noticed that several users of my game are sometimes experiencing bad performances. The game is generally running fine at my end. So I'm currently putting effort on this area and any low level optimization is always welcome for me 😉 Anyway I will get back to you if I can have a more concrete example useful for further investigations.

At the moment a "maybe" is better than a "no" 😄

**Comment by FliplineStudios** [comment]
Have there been any updates or workarounds for this issue?  We're starting to see what I'm guessing is something similar on one of our Windows 11 laptops, where back in 2023 it would have no issue with the framerate in our games on PC, but now (maybe after Windows Updates or driver updates) we're seeing very jerky performance when using the Nvidia GPU on the laptop.

For example, here's a snapshot from Scout.  When the game noticeably seems to be "losing" frames and suddenly getting jerky, we see the framerate spiking and dropping, with both "Waiting for GPU" and "Waiting for Next Frame" showing very high values suddenly:

![Image](https://github.com/user-attachments/assets/679babc7-c6d3-4d6f-a2c9-bfd104002d13)

You can see our actual frame time from the game (AS/DisplayList/etc.) is well below the red line for 60fps and never approaches it, but we see the jagged spikes and drops in the gray bars for "Waiting for GPU" and "Waiting for Next Frame".

This is happening on the Nvidia GeForce RTX 3060 Laptop GPU (which previously didn't have any issues with our games), running the game with DXD 9.  Same issues whether it's GPU or Direct render mode, same issues whether it's 32-bit or 64-bit.  We've tried versions back to 33.1.1.935 and up to 51.1.3.6, and same issue in all cases.

Strangely, if we switch to the built-in GPU of Intel Iris Xe Graphics, it doesn't seem to have any of these issues...  Not sure if this is specifically an Nvidia issue, we've been trying different versions of drivers and different settings in the "Manage 3D Settings" window, but nothing seems to make a difference.

**Comment by ajwfrost** [comment]
Hi @FliplineStudios  - I'm not 100% sure whether this is going to be just caused by how AIR works, or how it works in conjunction with certain drivers/hardware. Seems odd that it works okay with the built-in GPU; I don't suppose you have other (non-AIR) games/tests running that you could use also for comparison?

One update we have made which will come out in 51.2 is the use of ANGLE - basically, to allow us to use the OpenGL ES rendering mechanism which is in place for Android and iOS, which can also then work on Windows via this (Google-driven) intermediate layer. It's what Chrome uses to translate WebGL calls across the various platforms, and we're hoping this will help with stability/future-proofing across the different platforms.

The first 51.2 beta should come out very soon, it would interesting to see what you get from the performance of it!

thanks

**Comment by itlancer** [comment]
@ajwfrost, I can confirm that "Waiting for GPU" and "Waiting for Next Frame" very critical for every application across all platforms. Tested with multiple OSes (Windows, Android, Linux, iOS, macOS) with different devices, architechtures, OS versions, different GPUs (discrete, integrated). With `DisplayList` and with `Stage3D` render. With low-end and very high performance devices.
Doesn't matter - anyway to will get very laggy render when you try to see it in details. You can see stuttering almost in every where it should be smooth.
In all cases **direct** renderMode used. Not tested with others cause not actual in most of our cases.

For example: https://github.com/airsdk/Adobe-Runtime-Support/issues/3788
Even such sample without video (just rectangle rotation on each frame) sometimes cause little lags.
Seems framerate not stable. Maybe some internal AIR algorithm try to "compensate difference" between frames or something like that and "add" extra time for some frames. And as result instead of stable 60 FPS we get 58, then 62, then 58 etc. Even for applications without high load, with very simple scenarios.
In Scout it looks like that (gray "columns"):
![Image](https://github.com/user-attachments/assets/9984dff2-e607-45d7-bc4f-0abbefcd1291)

Moreover. Seems all works worse in simple scenarios (without render/logic overloading or extra frame time consumption). I mean if we try to optimize our high load applications - all render start to looks more "jerky". It makes me think that something wrong with AIR "enterFrame/frame time" or something like that. Framerate not stable at all.

Also to reproduce such issues you can use simple samples:
1) Simple camera output with `Stage3D`/`VideoTexture`: https://github.com/airsdk/Adobe-Runtime-Support/issues/3762
2) Simple camera output with `DisplayList`/`Video`: https://github.com/airsdk/Adobe-Runtime-Support/issues/2693

In real cases all these much worse sometimes with "no-reason" lags (it's not GC):
![Image](https://github.com/user-attachments/assets/c4b3f56c-e16d-495f-a0c2-6a2496438066)

Maybe related with:
https://github.com/airsdk/Adobe-Runtime-Support/issues/1125
https://github.com/airsdk/Adobe-Runtime-Support/issues/3616

**Comment by hardcoremore** [comment]
I can confirm this also:

![Image](https://github.com/user-attachments/assets/d4e9f2c0-dc27-4134-8232-dd493e322477)

If you look a the screenshot you can see that it waits 17ms for next frame which suggest that something is very wrong with calculations for next frame.

Other things I noticed is that game was running much more smooth when Scout was not opened. But I was testing release build here without advanced telemetry so I don't know how that can interrupt results.

This was tested on 51.1.3 since 51.2 does not work for me. it was windows 11 also
